File: python/advanced_ds/persistent_segment_tree.py
"""
Persistent Segment Tree (Structural Sharing)

Time Complexity:
- Update: O(log n) per version
- Query: O(log n) per query
- Space Complexity: O((n + m) * log n) where m = number of updates

Use Cases:
- Query array state at any previous version
- Historical queries on time-versioned data
- Undo/redo functionality
- Range queries with version history
- Counting distinct elements in range (via persistent data structures)

Key Insight:
- Use structural sharing to avoid duplicating entire trees per update
- Only create new nodes along the path modified during update
- Reuse unmodified subtrees from previous versions
- Allows O(log n) access to any previous version
"""

from typing import List, Optional, Tuple


class PersistentSegmentTree:
    """Persistent segment tree using structural sharing."""

    class Node:
        """Segment tree node with optional children."""

        def __init__(self, value: int = 0):
            self.value = value
            self.left = None
            self.right = None

        def copy(self):
            """Create a shallow copy of this node."""
            new_node = PersistentSegmentTree.Node(self.value)
            new_node.left = self.left
            new_node.right = self.right
            return new_node

    def __init__(self, arr: List[int]):
        """Initialize persistent segment tree."""
        self.n = len(arr)
        self.versions = []  # Store root of each version

        if self.n > 0:
            root = self._build(arr, 0, self.n - 1)
            self.versions.append(root)

    def _build(self, arr: List[int], start: int, end: int) -> Optional['PersistentSegmentTree.Node']:
        """Build initial tree."""
        if start == end:
            return self.Node(arr[start])

        mid = (start + end) // 2
        node = self.Node()
        node.left = self._build(arr, start, mid)
        node.right = self._build(arr, mid + 1, end)
        node.value = node.left.value + node.right.value
        return node

    def _update(self, node: Optional['PersistentSegmentTree.Node'], start: int, end: int,
                idx: int, val: int) -> Optional['PersistentSegmentTree.Node']:
        """Update with structural sharing."""
        if node is None:
            return None

        if start == end:
            # Create new leaf node
            return self.Node(val)

        # Copy current node
        new_node = node.copy()
        mid = (start + end) // 2

        if idx <= mid:
            # Update left subtree
            new_node.left = self._update(node.left, start, mid, idx, val)
        else:
            # Update right subtree
            new_node.right = self._update(node.right, mid + 1, end, idx, val)

        # Recompute value
        left_val = new_node.left.value if new_node.left else 0
        right_val = new_node.right.value if new_node.right else 0
        new_node.value = left_val + right_val

        return new_node

    def update(self, idx: int, val: int) -> int:
        """
        Create new version with update at index idx to val.

        Returns:
            Version number (0-indexed)
        """
        if not self.versions:
            return -1

        new_root = self._update(self.versions[-1], 0, self.n - 1, idx, val)
        self.versions.append(new_root)
        return len(self.versions) - 1

    def _query(self, node: Optional['PersistentSegmentTree.Node'], start: int, end: int,
               l: int, r: int) -> int:
        """Query sum in range [l, r]."""
        if node is None or start > r or end < l:
            return 0

        if l <= start and end <= r:
            return node.value

        mid = (start + end) // 2
        left_sum = self._query(node.left, start, mid, l, r)
        right_sum = self._query(node.right, mid + 1, end, l, r)
        return left_sum + right_sum

    def query(self, version: int, l: int, r: int) -> int:
        """
        Query sum in range [l, r] for a specific version.

        Args:
            version: Version number
            l: Left index (inclusive)
            r: Right index (inclusive)

        Returns:
            Sum of elements in range for that version
        """
        if version < 0 or version >= len(self.versions):
            return 0

        return self._query(self.versions[version], 0, self.n - 1, l, r)

    def point_query(self, version: int, idx: int) -> int:
        """Query value at index idx for a specific version."""
        return self.query(version, idx, idx)

    def get_num_versions(self) -> int:
        """Get number of versions."""
        return len(self.versions)


class PersistentSegmentTreeWithMax:
    """Persistent segment tree for range max queries."""

    class Node:
        """Segment tree node."""

        def __init__(self, value: int = float('-inf')):
            self.value = value
            self.left = None
            self.right = None

        def copy(self):
            """Create a shallow copy."""
            new_node = PersistentSegmentTreeWithMax.Node(self.value)
            new_node.left = self.left
            new_node.right = self.right
            return new_node

    def __init__(self, arr: List[int]):
        """Initialize persistent segment tree."""
        self.n = len(arr)
        self.versions = []

        if self.n > 0:
            root = self._build(arr, 0, self.n - 1)
            self.versions.append(root)

    def _build(self, arr: List[int], start: int, end: int) -> Optional['PersistentSegmentTreeWithMax.Node']:
        """Build initial tree."""
        if start == end:
            return self.Node(arr[start])

        mid = (start + end) // 2
        node = self.Node()
        node.left = self._build(arr, start, mid)
        node.right = self._build(arr, mid + 1, end)
        node.value = max(node.left.value, node.right.value)
        return node

    def _update(self, node: Optional['PersistentSegmentTreeWithMax.Node'], start: int, end: int,
                idx: int, val: int) -> Optional['PersistentSegmentTreeWithMax.Node']:
        """Update with structural sharing."""
        if node is None:
            return None

        if start == end:
            return self.Node(val)

        new_node = node.copy()
        mid = (start + end) // 2

        if idx <= mid:
            new_node.left = self._update(node.left, start, mid, idx, val)
        else:
            new_node.right = self._update(node.right, mid + 1, end, idx, val)

        left_val = new_node.left.value if new_node.left else float('-inf')
        right_val = new_node.right.value if new_node.right else float('-inf')
        new_node.value = max(left_val, right_val)

        return new_node

    def update(self, idx: int, val: int) -> int:
        """Create new version with update. Returns version number."""
        if not self.versions:
            return -1

        new_root = self._update(self.versions[-1], 0, self.n - 1, idx, val)
        self.versions.append(new_root)
        return len(self.versions) - 1

    def _query_max(self, node: Optional['PersistentSegmentTreeWithMax.Node'], start: int, end: int,
                   l: int, r: int) -> int:
        """Query max in range."""
        if node is None or start > r or end < l:
            return float('-inf')

        if l <= start and end <= r:
            return node.value

        mid = (start + end) // 2
        left_max = self._query_max(node.left, start, mid, l, r)
        right_max = self._query_max(node.right, mid + 1, end, l, r)
        return max(left_max, right_max)

    def query_max(self, version: int, l: int, r: int) -> int:
        """Query max in range [l, r] for specific version."""
        if version < 0 or version >= len(self.versions):
            return float('-inf')

        return self._query_max(self.versions[version], 0, self.n - 1, l, r)

    def get_num_versions(self) -> int:
        """Get number of versions."""
        return len(self.versions)


if __name__ == "__main__":
    # Example 1: Range sum with version history
    print("=== Range Sum with Version History ===")
    arr = [1, 2, 3, 4, 5]
    pst = PersistentSegmentTree(arr)

    print(f"Initial array: {arr}")
    print(f"Version 0, Query [0, 4]: {pst.query(0, 0, 4)}")  # 15

    v1 = pst.update(2, 10)  # Change arr[2] from 3 to 10
    print(f"\nAfter update(2, 10):")
    print(f"Version 0, Query [0, 4]: {pst.query(0, 0, 4)}")  # 15 (unchanged)
    print(f"Version {v1}, Query [0, 4]: {pst.query(v1, 0, 4)}")  # 22

    v2 = pst.update(0, 100)  # Change arr[0] from 1 to 100
    print(f"\nAfter update(0, 100):")
    print(f"Version 0, Query [0, 4]: {pst.query(0, 0, 4)}")  # 15 (unchanged)
    print(f"Version {v1}, Query [0, 4]: {pst.query(v1, 0, 4)}")  # 22 (unchanged)
    print(f"Version {v2}, Query [0, 4]: {pst.query(v2, 0, 4)}")  # 122

    # Example 2: Range max with version history
    print("\n=== Range Max with Version History ===")
    arr2 = [3, 1, 4, 1, 5, 9, 2, 6]
    pst2 = PersistentSegmentTreeWithMax(arr2)

    print(f"Initial array: {arr2}")
    print(f"Version 0, Query max [0, 7]: {pst2.query_max(0, 0, 7)}")  # 9

    v3 = pst2.update(1, 100)
    print(f"\nAfter update(1, 100):")
    print(f"Version 0, Query max [0, 7]: {pst2.query_max(0, 0, 7)}")  # 9
    print(f"Version {v3}, Query max [0, 7]: {pst2.query_max(v3, 0, 7)}")  # 100
    print(f"Version {v3}, Query max [0, 3]: {pst2.query_max(v3, 0, 3)}")  # 100

File: java/advanced_ds/PersistentSegmentTree.java
package advanced_ds;

import java.util.*;

/**
 * Persistent Segment Tree (Structural Sharing)
 *
 * Time Complexity:
 * - Update: O(log n) per version
 * - Query: O(log n) per query
 * - Space Complexity: O((n + m) * log n) where m = number of updates
 *
 * Use Cases:
 * - Query array state at any previous version
 * - Historical queries on time-versioned data
 * - Undo/redo functionality
 * - Range queries with version history
 * - Counting distinct elements in range (via persistent data structures)
 *
 * Key Insight:
 * - Use structural sharing to avoid duplicating entire trees per update
 * - Only create new nodes along the path modified during update
 * - Reuse unmodified subtrees from previous versions
 * - Allows O(log n) access to any previous version
 */
public class PersistentSegmentTree {

    /**
     * Node for range sum version.
     */
    static class Node {
        long value;
        Node left, right;

        Node(long value) {
            this.value = value;
        }

        Node copy() {
            Node newNode = new Node(this.value);
            newNode.left = this.left;
            newNode.right = this.right;
            return newNode;
        }
    }

    private int n;
    private List<Node> versions;

    /**
     * Initialize persistent segment tree.
     *
     * @param arr Initial array
     */
    public PersistentSegmentTree(int[] arr) {
        this.n = arr.length;
        this.versions = new ArrayList<>();

        if (n > 0) {
            Node root = build(arr, 0, n - 1);
            versions.add(root);
        }
    }

    private Node build(int[] arr, int start, int end) {
        if (start == end) {
            return new Node(arr[start]);
        }

        int mid = (start + end) / 2;
        Node node = new Node(0);
        node.left = build(arr, start, mid);
        node.right = build(arr, mid + 1, end);
        node.value = node.left.value + node.right.value;
        return node;
    }

    private Node update(Node node, int start, int end, int idx, long val) {
        if (node == null) {
            return null;
        }

        if (start == end) {
            return new Node(val);
        }

        Node newNode = node.copy();
        int mid = (start + end) / 2;

        if (idx <= mid) {
            newNode.left = update(node.left, start, mid, idx, val);
        } else {
            newNode.right = update(node.right, mid + 1, end, idx, val);
        }

        long leftVal = newNode.left != null ? newNode.left.value : 0;
        long rightVal = newNode.right != null ? newNode.right.value : 0;
        newNode.value = leftVal + rightVal;

        return newNode;
    }

    /**
     * Create new version with update at index idx to val.
     *
     * @param idx Index to update
     * @param val New value
     * @return Version number (0-indexed)
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public int update(int idx, long val) {
        if (versions.isEmpty()) {
            return -1;
        }

        Node newRoot = update(versions.get(versions.size() - 1), 0, n - 1, idx, val);
        versions.add(newRoot);
        return versions.size() - 1;
    }

    private long query(Node node, int start, int end, int l, int r) {
        if (node == null || start > r || end < l) {
            return 0;
        }

        if (l <= start && end <= r) {
            return node.value;
        }

        int mid = (start + end) / 2;
        long leftSum = query(node.left, start, mid, l, r);
        long rightSum = query(node.right, mid + 1, end, l, r);
        return leftSum + rightSum;
    }

    /**
     * Query sum in range [l, r] for a specific version.
     *
     * @param version Version number
     * @param l Left index (inclusive)
     * @param r Right index (inclusive)
     * @return Sum of elements in range
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public long query(int version, int l, int r) {
        if (version < 0 || version >= versions.size()) {
            return 0;
        }

        return query(versions.get(version), 0, n - 1, l, r);
    }

    /**
     * Query value at index idx for a specific version.
     *
     * @param version Version number
     * @param idx Index
     * @return Value at index
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public long pointQuery(int version, int idx) {
        return query(version, idx, idx);
    }

    /**
     * Get number of versions.
     *
     * @return Number of versions
     */
    public int getNumVersions() {
        return versions.size();
    }

    /**
     * Node for range max version.
     */
    static class MaxNode {
        long value;
        MaxNode left, right;

        MaxNode(long value) {
            this.value = value;
        }

        MaxNode copy() {
            MaxNode newNode = new MaxNode(this.value);
            newNode.left = this.left;
            newNode.right = this.right;
            return newNode;
        }
    }

    /**
     * Persistent segment tree for range max queries.
     */
    static class RangeMaxTree {
        private int n;
        private List<MaxNode> versions;

        RangeMaxTree(int[] arr) {
            this.n = arr.length;
            this.versions = new ArrayList<>();

            if (n > 0) {
                MaxNode root = build(arr, 0, n - 1);
                versions.add(root);
            }
        }

        private MaxNode build(int[] arr, int start, int end) {
            if (start == end) {
                return new MaxNode(arr[start]);
            }

            int mid = (start + end) / 2;
            MaxNode node = new MaxNode(Long.MIN_VALUE);
            node.left = build(arr, start, mid);
            node.right = build(arr, mid + 1, end);
            node.value = Math.max(node.left.value, node.right.value);
            return node;
        }

        private MaxNode update(MaxNode node, int start, int end, int idx, long val) {
            if (node == null) {
                return null;
            }

            if (start == end) {
                return new MaxNode(val);
            }

            MaxNode newNode = node.copy();
            int mid = (start + end) / 2;

            if (idx <= mid) {
                newNode.left = update(node.left, start, mid, idx, val);
            } else {
                newNode.right = update(node.right, mid + 1, end, idx, val);
            }

            long leftVal = newNode.left != null ? newNode.left.value : Long.MIN_VALUE;
            long rightVal = newNode.right != null ? newNode.right.value : Long.MIN_VALUE;
            newNode.value = Math.max(leftVal, rightVal);

            return newNode;
        }

        /**
         * [Brief description]
         *
         * @param [param] [description]
         * @return [description]
         * @time O([complexity])
         */
        public int update(int idx, long val) {
            if (versions.isEmpty()) {
                return -1;
            }

            MaxNode newRoot = update(versions.get(versions.size() - 1), 0, n - 1, idx, val);
            versions.add(newRoot);
            return versions.size() - 1;
        }

        private long queryMax(MaxNode node, int start, int end, int l, int r) {
            if (node == null || start > r || end < l) {
                return Long.MIN_VALUE;
            }

            if (l <= start && end <= r) {
                return node.value;
            }

            int mid = (start + end) / 2;
            long leftMax = queryMax(node.left, start, mid, l, r);
            long rightMax = queryMax(node.right, mid + 1, end, l, r);
            return Math.max(leftMax, rightMax);
        }

        /**
         * [Brief description]
         *
         * @param [param] [description]
         * @return [description]
         * @time O([complexity])
         */
        public long queryMax(int version, int l, int r) {
            if (version < 0 || version >= versions.size()) {
                return Long.MIN_VALUE;
            }

            return queryMax(versions.get(version), 0, n - 1, l, r);
        }

        /**
         * [Brief description]
         *
         * @param [param] [description]
         * @return [description]
         * @time O([complexity])
         */
        public int getNumVersions() {
            return versions.size();
        }
    }

    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static void main(String[] args) {
        // Example 1: Range sum with version history
        System.out.println("=== Range Sum with Version History ===");
        int[] arr = {1, 2, 3, 4, 5};
        PersistentSegmentTree pst = new PersistentSegmentTree(arr);

        System.out.println("Initial array: [1, 2, 3, 4, 5]");
        System.out.println("Version 0, Query [0, 4]: " + pst.query(0, 0, 4));  // 15

        int v1 = pst.update(2, 10);  // Change arr[2] from 3 to 10
        System.out.println("\nAfter update(2, 10):");
        System.out.println("Version 0, Query [0, 4]: " + pst.query(0, 0, 4));  // 15 (unchanged)
        System.out.println("Version " + v1 + ", Query [0, 4]: " + pst.query(v1, 0, 4));  // 22

        int v2 = pst.update(0, 100);  // Change arr[0] from 1 to 100
        System.out.println("\nAfter update(0, 100):");
        System.out.println("Version 0, Query [0, 4]: " + pst.query(0, 0, 4));  // 15 (unchanged)
        System.out.println("Version " + v1 + ", Query [0, 4]: " + pst.query(v1, 0, 4));  // 22 (unchanged)
        System.out.println("Version " + v2 + ", Query [0, 4]: " + pst.query(v2, 0, 4));  // 122

        // Example 2: Range max with version history
        System.out.println("\n=== Range Max with Version History ===");
        int[] arr2 = {3, 1, 4, 1, 5, 9, 2, 6};
        RangeMaxTree pst2 = new RangeMaxTree(arr2);

        System.out.println("Initial array: [3, 1, 4, 1, 5, 9, 2, 6]");
        System.out.println("Version 0, Query max [0, 7]: " + pst2.queryMax(0, 0, 7));  // 9

        int v3 = pst2.update(1, 100);
        System.out.println("\nAfter update(1, 100):");
        System.out.println("Version 0, Query max [0, 7]: " + pst2.queryMax(0, 0, 7));  // 9
        System.out.println("Version " + v3 + ", Query max [0, 7]: " + pst2.queryMax(v3, 0, 7));  // 100
        System.out.println("Version " + v3 + ", Query max [0, 3]: " + pst2.queryMax(v3, 0, 3));  // 100
    }
}
